data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamScheduler.java

```diff
@@ -0,0 +1,85 @@
+package org.opensearch.dataprepper.plugins.mongo.stream;
+
+import org.opensearch.dataprepper.buffer.common.BufferAccumulator;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourceCoordinator;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourcePartition;
+import org.opensearch.dataprepper.plugins.mongo.buffer.ExportRecordBufferWriter;
+import org.opensearch.dataprepper.plugins.mongo.buffer.RecordBufferWriter;
+import org.opensearch.dataprepper.plugins.mongo.configuration.MongoDBSourceConfig;
+import org.opensearch.dataprepper.plugins.mongo.converter.RecordConverter;
+import org.opensearch.dataprepper.plugins.mongo.coordination.partition.StreamPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+public class StreamScheduler implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(StreamScheduler.class);
+    private static final int DEFAULT_TAKE_LEASE_INTERVAL_MILLIS = 60_000;
+    /**
+     * Number of records to accumulate before flushing to buffer
+     */
+    static final int DEFAULT_BUFFER_BATCH_SIZE = 10;
```

Comment:
I originally thought that this could lead to high latency in test scenarios, but we are already flushing the buffer every time we write to it here (https://github.com/opensearch-project/data-prepper/blob/8760945c8091d1dd4e2bb8c66c5bb9f33f0aaa43/data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/buffer/ExportRecordBufferWriter.java#L76)